File: src/codex_subagent_kit/usage.py
from __future__ import annotations

from pathlib import Path

from .catalog import get_agents
from .generator import DEFAULT_ORCHESTRATOR_KEY, ORCHESTRATOR_CATEGORY


class UsageError(RuntimeError):
    """Raised when usage guidance cannot be rendered."""


def _visible_installed_agents(*, project_root: Path, scope: str):
    include_project = scope == "project"
    agents = get_agents(
        project_root=project_root,
        include_project=include_project,
        include_global=True,
    )
    visible = [agent for agent in agents if agent.source in {"project", "global"}]
    if not visible:
        raise UsageError("no installed agents were found for the selected scope")
    return visible


def _starter_prompt(*, task: str, orchestrator_key: str | None, worker_keys: list[str]) -> str:
    if orchestrator_key:
        if worker_keys:
            worker_list = ", ".join(worker_keys)
            return (
                f'Use {orchestrator_key} as the root orchestrator for this task: "{task}". '
                f"Delegate to {worker_list} when appropriate."
            )
        return f'Use {orchestrator_key} for this task: "{task}".'

    if worker_keys:
        worker_list = ", ".join(worker_keys)
        return f'For this task: "{task}", use these installed agents as needed: {worker_list}.'

    raise UsageError("no installed agents were found for the selected scope")


def render_usage_guide(*, project_root: Path, scope: str, task: str | None = None) -> str:
    visible_agents = _visible_installed_agents(project_root=project_root, scope=scope)
    task_text = (task or "<describe the task here>").strip()

    orchestrators = [agent for agent in visible_agents if agent.category == ORCHESTRATOR_CATEGORY]
    orchestrator_key: str | None = None
    if orchestrators:
        preferred = next((agent for agent in orchestrators if agent.key == DEFAULT_ORCHESTRATOR_KEY), None)
        orchestrator_key = (preferred or orchestrators[0]).key

    worker_keys = [agent.key for agent in visible_agents if agent.key != orchestrator_key]

    lines = [
        f"scope: {scope}",
        "visible installed agents:",
    ]
    for agent in visible_agents:
        lines.append(f"- {agent.key} [{agent.source}] - {agent.description}")

    lines.extend(
        [
            "",
            "starter prompt:",
            _starter_prompt(
                task=task_text,
                orchestrator_key=orchestrator_key,
                worker_keys=worker_keys,
            ),
            "",
            "direct prompt ideas:",
        ]
    )
    for agent in visible_agents[:6]:
        lines.append(f'- Use {agent.key} for this task: "{task_text}". Focus on: {agent.description}')

    lines.extend(
        [
            "",
            "session tip:",
            "- If Codex spawns a subagent thread, use /agent to inspect or continue that thread.",
        ]
    )
    return "\n".join(lines)

File: packages/codex-subagent-kit/src/usage.ts
import { getAgents } from "./catalog";
import { DEFAULT_ORCHESTRATOR_KEY, ORCHESTRATOR_CATEGORY } from "./generator";

export class UsageError extends Error {}

export interface RenderUsageGuideOptions {
  projectRoot: string;
  scope: "project" | "global";
  task?: string;
  homeDir?: string;
}

function visibleInstalledAgents(options: RenderUsageGuideOptions) {
  const agents = getAgents({
    projectRoot: options.projectRoot,
    homeDir: options.homeDir,
    includeProject: options.scope === "project",
    includeGlobal: true,
  });
  const visible = agents.filter((agent) => agent.source === "project" || agent.source === "global");
  if (visible.length === 0) {
    throw new UsageError("no installed agents were found for the selected scope");
  }
  return visible;
}

function starterPrompt(task: string, orchestratorKey: string | undefined, workerKeys: string[]): string {
  if (orchestratorKey) {
    if (workerKeys.length > 0) {
      return `Use ${orchestratorKey} to coordinate this task: "${task}". Delegate to ${workerKeys.join(", ")} when appropriate.`;
    }
    return `Use ${orchestratorKey} for this task: "${task}".`;
  }

  if (workerKeys.length > 0) {
    return `For this task: "${task}", use these installed agents as needed: ${workerKeys.join(", ")}.`;
  }

  throw new UsageError("no installed agents were found for the selected scope");
}

export function renderUsageGuide(options: RenderUsageGuideOptions): string {
  const visibleAgents = visibleInstalledAgents(options);
  const taskText = (options.task ?? "<describe the task here>").trim();

  const orchestrators = visibleAgents.filter((agent) => agent.category === ORCHESTRATOR_CATEGORY);
  let orchestratorKey: string | undefined;
  if (orchestrators.length > 0) {
    orchestratorKey =
      orchestrators.find((agent) => agent.key === DEFAULT_ORCHESTRATOR_KEY)?.key ??
      orchestrators[0]?.key;
  }

  const workerKeys = visibleAgents
    .filter((agent) => agent.key !== orchestratorKey)
    .map((agent) => agent.key);

  const lines = [
    `scope: ${options.scope}`,
    "visible installed agents:",
    ...visibleAgents.map((agent) => `- ${agent.key} [${agent.source}] - ${agent.description}`),
    "",
    "starter prompt:",
    starterPrompt(taskText, orchestratorKey, workerKeys),
    "",
    "direct prompt ideas:",
    ...visibleAgents.slice(0, 6).map(
      (agent) => `- Use ${agent.key} for this task: "${taskText}". Focus on: ${agent.description}`,
    ),
    "",
    "session tip:",
    "- If Codex spawns a subagent thread, use /agent to inspect or continue that thread.",
  ];

  return lines.join("\n");
}
